Enable use with `python -m` or `pythonw -m`

MHendricks opened the pull request:
Add support for `python -m pyqt6_tools`. The ultimate goal is to on windows be able to create a desktop shortcut that opens QtDesigner without showing a command prompt that has to stay open for the duration that designer is running. This is accomplished by launching using `pythonw.exe` instead of `python.exe` instead of the console_scripts entry_point.

An alternative way to accomplish this could be to add a second `gui_scripts` exe to setup.py. This ends up adding a second executable, I'm not sure if this is desirable.
setup.py:
```py
    entry_points={
        'console_scripts': [
            'pyqt{major}-tools = pyqt{major}_tools.entrypoints:main'.format(major=pyqt_major_version),
        ],
        'gui_scripts': [
            'pyqt{major}-toolsw = pyqt{major}_tools.entrypoints:main'.format(major=pyqt_major_version),
        ]
    }
```

Also, I'm still needing to support PyQt5 so I would want to make sure this is added to the pyqt5-tools releases as well.

altendky commented on src/pyqt_tools/__main__.py:
```diff
@@ -0,0 +1,19 @@
+import pathlib
+import sys
+
+if __name__ == "__main__":
```
I would personally drop this since there's no point in importing this for any other purpose and modules that you can both run and import can result in really goofy corner cases when you do happen to do both of those things in one process.

MHendricks replied:
True, it doesn't make sense for a `__main__.py` file and leads to the issues in the other note, removed.

altendky commented on src/pyqt_tools/__main__.py:
```diff
@@ -0,0 +1,19 @@
+import sys
+
+if __name__ == "__main__":
+    # Relative imports don't work when called by `python -m ...`, get the name
```
I'm confused.  No package that uses relative imports can be run as a module?  That doesn't sound right.

MHendricks replied:
Ah, you are right, I'm thinking of when you try to pass a .py file to python.
```
C:\Users\example>py "C:\Program Files\Python39\Lib\site-packages\pyqt5_tools\__main__.py"
Traceback (most recent call last):
  File "C:\Program Files\Python39\Lib\site-packages\pyqt5_tools\__main__.py", line 10, in <module>
    from . import entrypoints
ImportError: attempted relative import with no known parent package
```

That simplifies this code a lot. I copied the boilerplate for this from some of my own `__main__.py` files, which I'll probably revisit in the near future.

altendky reviewed (changes requested):
Hey, thanks for your interest and taking the time to make a PR.  I generally am very much in favor of being able to use `-m` so I'm all for this addition.  I'm a bit disappointed that I hadn't already added it...

MHendricks commented:
I'm happy with the changes at this point and it could be merged, but figured I would ask a few questions.

1. Is this able to be included in a PyQt5-tools release, I'm going to be needing it for that one.
2. Should I look into adding testing? I haven't dug into the testing setup for this yet so I don't know where to add tests for this. I'm not sure a test is strictly needed for a simple use case like this. I also notice a few tests are failing but that seems unrelated to my changes.
3. I'm guessing you don't want to add the optional gui_script `pyqt6-toolsw.exe` idea. That seems redundant and would add a exe to other platforms where it's not needed and I'm not sure if pythonw is a convention on the other platforms.
4. This is outside of the scope of the Pull Request, but this PR enables it so I'll mention it. When using this with `pythonw.exe -m` (from a shortcut so there is no host command prompt) you may notice a flickering of several Command Prompt windows opening and closing quickly. This is caused by the various calls to `subprocess.Popen/run/call` with default arguments. This is a minor issue and you may not see it unless you are cursed to be looking for it like I am 😄 .  The fix to this likely is to add `creationflags=subprocess.CREATE_NO_WINDOW` to all subprocess calls made.

altendky commented:
Not specifically related, but generally related to contributions and getting them into releases...  I have been bad about releasing because every time I come back to this project (which I haven't personally used for a couple years now) to create a new release I end up in some sort of ci/build mess.  I _think_ I ended up figuring out the last issue while working on another project, but didn't get back here to make sure and get the release finished.  So, maybe your contribution ends up kick starting a new attempt at a release, but be aware of the state of maintenance on this project, if you aren't already.

I like all your ideas.  Of the list here I think only the test addition belongs in this PR.  We can finish that and merge this and you can offer the other changes in other PRs.

> 1. Is this able to be included in a PyQt5-tools release, I'm going to be needing it for that one.

The package doesn't have a version in the name anymore because it is used to create both 6 and 5.  Pending the above commentary on project state, I would expect any changes to be available for both 5 and 6.  Albeit limited Qt versions of each.

> 2. Should I look into adding testing? I haven't dug into the testing setup for this yet so I don't know where to add tests for this. I'm not sure a test is strictly needed for a simple use case like this. I also notice a few tests are failing but that seems unrelated to my changes.

I rarely consider cases too simple to deserve a test.  I do sometimes not get around to adding tests.  It does seem appropriate to add this to this PR.  We could make [`executable_path_string`](https://github.com/altendky/pyqt-tools/blob/51fd15bc6d3223f087402a3a5a3218a4dcb4ebf7/src/pyqt_tools/tests/test_entrypoints.py#L26) a pytest fixture that results in both the present value and also a `-m` option.  I guess the name should change a bit and it should become a list that we unpack in the subprocess call so it can be either `[whatever_it_is_now]` or `[sys.executable, "-m", whatever_the_helper_is]`.

> 3. I'm guessing you don't want to add the optional gui_script `pyqt6-toolsw.exe` idea. That seems redundant and would add a exe to other platforms where it's not needed and I'm not sure if pythonw is a convention on the other platforms.

Go for it.

> 4. This is outside of the scope of the Pull Request, but this PR enables it so I'll mention it. When using this with `pythonw.exe -m` (from a shortcut so there is no host command prompt) you may notice a flickering of several Command Prompt windows opening and closing quickly. This is caused by the various calls to `subprocess.Popen/run/call` with default arguments. This is a minor issue and you may not see it unless you are cursed to be looking for it like I am 😄 .  The fix to this likely is to add `creationflags=subprocess.CREATE_NO_WINDOW` to all subprocess calls made.

I don't know offhand if that should be only conditionally set on Windows, but in general I'm good with addressing this.  I have had to fiddle around a bit with those creation flags in another project for other reasons.

MHendricks commented:
Ok I think I've addressed everything and this is good to go.

Tests have been added, your suggestion told me what I needed to know. I replaced each use of `executable_path_string` with the `pytest.mark.parametrize` decorator so the tests are run for each individual way to invoke it. I was a little hesitant to tackle the tests as I wasn't able to figure out how to run the tox setup in the ~15 minutes I gave to trying to figure out what ci.yaml was configuring. Luckily the pytest changes were simple enough that I only had to run the tests twice using a testing PR on my fork. I didn't see any guides on how to manually run the tests locally, that would be nice to have added to the documentation.

I also added the pyqt5-toolsw gui_scripts option, and was even able to include it in testing because you aren't capturing the text output in your tests.

altendky commented:
Take a look at [`5e178e9` (#123)](https://github.com/altendky/pyqt-tools/pull/123/commits/5e178e913013b456c9b586ca7c9f50047987d89a) for how to switch to a fixture instead of separate parametrization (though your unpacking was a nice touch with the separate parametrizations).

Other than that, I think we're good.  Thanks for the detailed effort on this contribution.

Yes, more docs could be good.  Given that you are the second contributor, if I remember the last 7 years or whatever correctly, I haven't been too worried.  Of course, maybe not making it easier on contributors is the reason...  `:]`  I'm not actually even sure the last time I ran the tests locally.

MHendricks commented:
That parameterized fixture feature is good to know about, makes sense for a reusable case like this. I've included your changes in this branch.

altendky reviewed (approved):
sorry for the wait...  thanks!
